### DISCUSSION TITLE
Copy a Shared album

### DISCUSSION THREAD

**Discussion by qdc** [xob0t/Google-Photos-Toolkit#31] [Ideas]
Hey, thanks for the tool, it looks promising.

I'm trying to find a workaround for changing the ownership of an album. 
The idea is to 
- A shares the album with B
- B saves all photos from the album
- B makes a copy of the album (with his own copies of the photos)
- A deletes the original album

I've been playing around with the photo APÏ for some time but 
a. I need to do a "hard copy" of each photo (download / upload), I cannot use the "Save photos" available in a shared album
b. Then add each photo in the correct order to a new album 
The API is limited to manipulating the photos that it created, hence this approach.
It's quite limited : I lose some metadata, and I cannot copy album enrichments (texts)? Preserving the album order is also tricky.

So I had the idea of using the web API and found your userscript. 

Would it be feasible to add the "Shared with me albums" in the album select box ? What do you think ? 
Thanks for your answer!

**Comment by qdc** [comment]
I've been trying to play with the API in the console, however with little success :

```
let nextPageId = null;
do {
  const page = await gptkApi.getSharedLinks(nextPageId);
  if (page == null) {break;}  // Add check to avoid error
  nextPageId = page.nextPageId;
  console.log("nextPageId");
} while (nextPageId);
console.log("DONE");
```

Doesn't return any meaningful data :
![Capture d'écran 2024-09-13 105805](https://github.com/user-attachments/assets/ecdbb87c-5b23-4b54-b66a-f63bfdec21df)


However when I navigate Google Photo and see the same request, the data for my Shared Albums is there :
![Capture d'écran 2024-09-13 105955](https://github.com/user-attachments/assets/813317ad-c020-4741-bf0d-2825e422746e)

So I figure it must be something with the following code :
```
  async getSharedLinks(pageId = null, parseResponse = true) {
    // type assertion
    if (pageId) assertType(pageId, 'string');
    if (parseResponse) assertType(parseResponse, 'boolean');

    const rpcid = 'F2A0H';
    const requestData = [pageId, null, 2, null, 3];
    try {
      const response = await this.makeApiRequest(rpcid, requestData);
      if (parseResponse) return parser(response, rpcid);
      return response;
    } catch (error) {
      console.error('Error in getSharedLinks:', error);
      throw error;
    }
  }
```

Maybe some requestData parameters to change... but I need to figure out their meaning

**Reply by xob0t** [reply]
I still haven't got time to reply properly, but it should be possible to do with current gptk api capabilites.
But not the enrichments part.
BTW, Shared links and shared albums are not the same thing. No way to get just shared albums, so use 'getAlbums', then filter them.

**Comment by qdc** [comment]
Thanks for your answer.  I can't get shared (with me) albums from getAlbums.

I actually managed to get them from sharedLinks by monitoring requests and responses from Google Photos and comparing them with the ones generated by GPTK API.
There is a difference in payload, which I fixed by changing the request data as follows.

`    const requestData = [pageId, null, 2, null, 2];`

Seems to work now, I'll keep you posted on my progress ;-)


Maybe this should be a different function named getSharedAlbums then ? I'm not so sure what I'm doing...
